lib/run-script-pkg.js

```diff
@@ -41,7 +45,7 @@ const runScriptPkg = async options => {
   )
     cmd = defaultGypInstallScript
   else if (event === 'start' && await isServerPackage(path))
-    cmd = 'node server.js' + args.map(a => ` ${JSON.stringify(a)}`).join('')
+    cmd = sh`node server.js ${args}`
```

Comment:
This could still have issues if `node` resolves to `node.exe` (and not some `node.cmd` or `node.bat` alias). `server.js` will probably see extra escaping sigils if `args` involves anything that needs escaping.

Reply:
👍 I’ll test this on a Windows machine soon

Reply:
I tested this. You are right: If `node` points to `node.exe`, the escaping is a problem.

I created `server.js` like this:

```js
console.log(process.argv)
```

And ran this from the Node.js REPL:

```
> require("./lib/run-script-pkg")({event:"start",path:process.cwd(),pkg:{},args:['"foo"']}).then(r=>console.log(r.stdout.toString()), console.error)
Promise { <pending> }
> [
  'C:\\Program Files\\nodejs\\node.exe',
  'C:\\Users\\Simon\\stuff\\run-script\\server.js',
  '^\\^foo\\^^'
]
```

As one can see, `server.js` didn’t receive `"foo"` as expected but `^\\^foo\\^^`.

However, if I install Node.js using `npm i node`, then `server.js` does receive `"foo"` since we go through npm’s batch script wrapper in `./node_modules/.bin/node.cmd`.

I fixed this by shipping with a tiny `node.cmd` wrapper that we can use always. With it, I get this in the REPL, both with `node.exe` and after `npm i node`:

```
> require("./lib/run-script-pkg")({event:"start",path:process.cwd(),pkg:{},args:['"foo"']}).then(r=>console.log(r.stdout.toString()), console.error)
Promise { <pending> }
> [
  'C:\\Program Files\\nodejs\\node.exe',
  'C:\\Users\\Simon\\stuff\\run-script\\server.js',
  '"foo"'
]
```

Here’s the fix: [`d1c4c26` (#31)](https://github.com/npm/run-script/pull/31/commits/d1c4c26e53f2e1b786cf24a36dd4f71880149c21)

Do you see any potential problems with this approach?

Reply:
Should work, although just to be safe maybe rename the wrapper to something like `nodewrapper.cmd`, or really anything other than `node.cmd`? If the directory containing the wrapper happened to be the working directory, or if it ever ended up in `%PATH%` somehow, you'd have an infinite loop on your hands.

Reply:
You could also do what `cross-spawn` does, which is to take a dependency on the `which` package, use that to find `node`, and check to see if it's an `.exe` or something else; and then use Puka's `quoteForWin32` or `quoteForCmd` respectively to escape the arguments. Or just use `cross-spawn` for the whole thing, assuming its escaping is as complete as Puka's (I have no reason to believe it is or isn't).